lib/ansible/modules/cloud/vmware/vmware_guest_customization.py

```diff
@@ -0,0 +1,563 @@
+        self.spec_name = self.params.get('spec_name')
+
+    def process_state(self):
+        spec_states = {
+            'absent': {
+                'present': self.state_remove_custom_spec,
+                'absent': self.state_exit_unchanged,
+            },
+            'present': {
+                'present': self.state_update_custom_spec,
+                'absent': self.state_create_custom_spec,
+            }
+        }
+        try:
+            spec_states[self.state][self.check_custom_spec_state()]()
+        except vmodl.RuntimeFault as runtime_fault:
+            self.module.fail_json(msg=to_native(runtime_fault.msg))
+        except vmodl.MethodFault as method_fault:
+            self.module.fail_json(msg=to_native(method_fault.msg))
+        except Exception as e:
+            self.module.fail_json(msg=to_native(e))
+
+    def check_custom_spec_state(self):
+        return 'present' if self.cc_mgr.DoesCustomizationSpecExist(name=self.spec_name) else 'absent'
+
+    def state_exit_unchanged(self):
+        self.module.exit_json(changed=False)
+
+    def state_remove_custom_spec(self):
+        if self.module.check_mode:
+            changed = True
+        else:
+            self.cc_mgr.DeleteCustomizationSpec(name=self.spec_name)
+            changed = True if self.check_custom_spec_state() == 'absent' else False
+
+        self.module.exit_json(changed=changed)
+
+    def state_update_custom_spec(self):
+        changed = False
+        current_spec = None
+        try:
+            current_spec = self.cc_mgr.GetCustomizationSpec(name=self.spec_name)
+        except vim.fault.NotFound as e:
+            self.module.fail_json(msg="Failed to retrieve customization"
+                                      " specification '%s' while updating : %s" % (self.spec_name,
+                                                                                   to_native(e.msg)))
+        spec_type = self.params.get('spec_type')
+        if spec_type != current_spec.info.type:
+            self.module.fail_json(msg="Can not change the type of customization"
+                                      " specification from '%s' to '%s'" % (spec_type,
+                                                                            current_spec.info.type))
+
+        new_spec_name = self.params.get('new_spec_name')
+        if new_spec_name and new_spec_name != current_spec.info.name:
+            if self.cc_mgr.DoesCustomizationSpecExist(name=new_spec_name):
+                self.module.fail_json(msg="Unable to rename customization specification. Please specify"
+                                          " different name for customization spec as '%s' already exists"
+                                          " in given configuration." % new_spec_name)
+            if not self.module.check_mode:
+                self.cc_mgr.RenameCustomizationSpec(name=self.spec_name, newName=new_spec_name)
+            self.module.exit_json(changed=True)
```

Comment:
You automatically exit on a rename, even while other values might change? Is that required?

Other way could be to do the rename at the end? After reconf has happened if that is necessary